### PR TITLE
[JENKINS-56229] Use the correct access token when impersonating a user

### PR DESCRIPTION
**Pull request by agentgonzo**
The token retrieved will be for the caller of the user.impersonate()
method (ie, could be SYSTEM), rather than that of the target of the
impersonation.

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java [hunk 1]
```diff
@@ -659,7 +659,7 @@ public UserDetails loadUserByUsername(String username)
 
         Authentication token = SecurityContextHolder.getContext().getAuthentication();
 
-        if (token == null) {
+        if (token == null || username != token.getPrincipal()) {
```
Comment (Wadeck) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]:
```suggestion
        if (token == null || !username.equals(token.getPrincipal())) {
```

String reference comparison could be problematic.

**Comment by jtnord** [comment]
@samrocketman are you still maintaining this plugin?

**Comment by alexanderrtaylor** [comment]
@samrocketman Is there anything we can do to help here? It is still showing you as the maintainer so I wanted to check if this could be approved and merged

**Comment by samrocketman** [comment]
Hello, I’ve been away a few months due to a cross country move.  I’m getting back into the open source groove.  I will be testing this change this week.

**Comment by Wadeck** [comment]
@samrocketman welcome back 🚢 ✈️ 🚌 🚗 🚴

**Comment by samrocketman** [comment]
This has been released in github-oauth 0.32

**Comment by agentgonzo** [comment]
Thanks @samrocketman !!!

**Comment by samrocketman** [comment]
Thanks for the fix.  I verified it worked by using a freestyle job and impersonating to a user.

```groovy
import jenkins.model.Jenkins

Jenkins.instance.as(User.get('samrocketman')).with { ctx ->
    try {
        // some code as user
    } finally {
        ctx.close()
    }
}
```

**Comment by agentgonzo** [comment]
>Before the fix it would throw an error that anonymous does not have read permissions. After, behaves as expected.

That was what we saw too

**Comment by samrocketman** [comment]
@agentgonzo mind taking a look at https://issues.jenkins-ci.org/browse/JENKINS-57154 ?

**Comment by samrocketman** [comment]
@Wadeck @agentgonzo 
I've discovered this PR is the root cause of https://issues.jenkins-ci.org/browse/JENKINS-57154

**Comment by samrocketman** [comment]
Reverting this change seems to fix things locally on my test machine.  Impersonation is still broken without it, though.

**Pull request merged**